cmd/api/insert.go

```diff
@@ -11,86 +14,94 @@ import (
-	if c.QueryBool("debug", false) {
-		rid := ulid.Make().String()
-		log.Debug().
-			Str("request_id", rid).
-			Interface("headers", c.GetReqHeaders()).
-			Str("body", string(c.Body())).
-			Interface("queryParams", c.Queries()).
-			Msg("Incoming request")
+	rid := ulid.Make().String()
+	log.Debug().
+		Str("request_id", rid).
+		Interface("headers", c.GetReqHeaders()).
+		Interface("queryParams", c.Queries()).
+		Msg("Incoming request")
+
+	body := c.Body()
+	if !gjson.ValidBytes(body) {
+		return fiber.NewError(http.StatusBadRequest, "invalid JSON")
 	}
 
 	// TODO: this block can be abstracted as we also use it for query
-	/////
 	apiKey := c.Locals("apiKey").(models.APIKey)
 
 	// TODO: read-only vs read-write connections
 	connectionSetting := a.db.GetDatabaseConnection(apiKey.DestinationID)
-
 	if connectionSetting.ID == "" {
-		return errors.New("No DB Connections set up")
+		return fiber.NewError(http.StatusUnauthorized, "no connection is set up")
 	}
-	/////
 
-	input := c.Body()
-	if !gjson.ValidBytes(input) {
-		return errors.New("Invalid JSON")
+	var (
+		tableName, flatAlgo string
+		fromBody            bool
+	)
+	flatAlgo = c.Get(FlattenHeader, c.Query(FlattenQuery))
+	tableName = c.Get(TableNameHeader, c.Query(TableNameQuery))
+	if tableName == "" {
+		if tableName = gjson.GetBytes(body, TableNameJson).String(); tableName == "" {
+			return fiber.NewError(http.StatusBadRequest, "missing required table field")
+		}
+		fromBody = true
 	}
 
-	// TODO: Use actual table name from request
-	tableName := "t"
-
-	var FlattenFunc = Flatten
-
-	parsed := gjson.ParseBytes(input)
-	if parsed.IsArray() {
-		for _, item := range parsed.Array() {
-			flat, err := FlattenFunc(item.Get(`@ugly`).Raw)
-			if err != nil {
-				log.Error().Err(err).Str("json", item.Str).Msg("Unable to flatten json")
-			}
-			a.dataTransport.Write(connectionSetting.ID, tableName, []byte(flat))
+	parsed := gjson.ParseBytes(body)
+	if fromBody {
+		if parsed = parsed.Get("data"); !parsed.Exists() {
+			return fiber.NewError(http.StatusBadRequest, "missing required data field")
 		}
+	}
+
+	var items []gjson.Result
+	if flatAlgo == "explode" && parsed.IsArray() {
+		items = append(items, parsed.Array()...)
 	} else {
-		flat, err := FlattenFunc(gjson.GetBytes(input, `@ugly`).Raw)
+		items = append(items, gjson.GetBytes(body, `@ugly`))
+	}
+
+	var err error
+	for _, item := range items {
+		flat, flattenErr := flatten.FlattenString(
+			item.Get(`@ugly`).Raw,
+			"",
+			flatten.UnderscoreStyle,
+		)
 		if err != nil {
-			return err
+			log.Error().Err(err).Str("json", item.Str).Msg("Unable to flatten json")
+			err = errors.Join(err, flattenErr)
+			continue
 		}
-		a.dataTransport.Write(connectionSetting.ID, tableName, []byte(flat))
+		writeErr := a.dataTransport.Write(connectionSetting.ID, tableName, []byte(flat))
+		if writeErr != nil {
+			err = errors.Join(err, flattenErr)
+		}
+	}
+	if err != nil {
+		return fiber.NewError(http.StatusMultiStatus, err.Error())
```

Comment:
We should probably return a non-2xx error here.

Reply:
@poundifdef Which status code would you recommend here?

Reply:
Used `http.StatusExpectationFailed: 417`